### DISCUSSION TITLE
Add support for DISTINCT and HAVING

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#2989] [Feature Requests]
**DISTINCT** is useful for getting unique results.

```sql
SELECT DISTINCT country FROM customers
SELECT COUNT(DISTINCT country) FROM customers
```

**HAVING** is used to filter against aggregate functions (like `WHERE` for grouped data).

```sql
SELECT COUNT(id), country
FROM customers
GROUP BY country
HAVING COUNT(id) > 5;
```

---

_Edit by @rijkvanzanten 8/24/2022: Directus now supports Aggregate functions and Grouping, so I've removed those from this feature request_

**Comment by EpariNigam** [comment]
How to add this functionality or what to use for this feature?

**Reply by rijkvanzanten** [reply]
I don't understand.. Are you asking how to build this or how to use it?

**Reply by EpariNigam** [reply]
Sorry for the misunderstanding, my question is how to use it in API?

**Reply by benhaynes** [reply]
It's a Discussion (Feature Request) and therefore doesn't exist yet.

**Comment by arladmin** [comment]
Hi there. Is there already an ETA for this feature?

**Reply by benhaynes** [reply]
[**Unfortunately, we can not provide estimates on features or fixes**](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

**Comment by benhaynes** [comment]
"How to fetch data by custom query group by" @nileshpande
---

As we can fetch distinct data from Mysql query by using group by simillarly we should be able to fetch data from directus api.
Either by using custom query parameter of any filter property.

ex:-https://abcd.com/abccar-api/items/cars

here are many records of cars. 
make: ford , Model: explorer
make: ford , Model: F-150
make: Honda, Model: civic

here I want to fetch distinct car makes

ex: ford,honda

it should not give ford data again.

If this already exists then please share any documentation with example.

**Reply by fitu86** [reply]
I do it this way, but grouping would be better.

![SVID_20220709_220427_1](https://user-images.githubusercontent.com/16139127/178129967-8c66bb09-af3d-4b1a-aac0-8fe474ac758b.gif)


https://m.youtube.com/watch?v=Fe9xP4VclME&feature=youtu.be

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by martin-braun** [reply]
Hi @rijkvanzanten, sorry if this is not related, but I kinda have problems with `groupBy` on nested fields. Essentially, I need this: https://github.com/directus/directus/discussions/15185

It's incredibly frustrating that feature requests get closed for not reaching the necessary vote count, because now I cannot vote on this (although I need it) and it prevents further discussion (i.e. how to bypass the limitation).

Could you please give advice how I can group by a nested field?

**Reply by rijkvanzanten** [reply]
> It's incredibly frustrating that feature requests get closed for not reaching the necessary vote count, because now I cannot vote on this (although I need it) and it prevents further discussion (i.e. how to bypass the limitation).

Happy to take alternative suggestions. The problem that it solves is that we ended up with a feature request backlog filled with noise that's flooding the list and therefore hiding the useful ones. I don't disagree there's some good ideas in there that don't make the cut, but so far it's been the best approach to keep the signal-to-noise-ratio in check 👍🏻 

> Could you please give advice how I can group by a nested field?

There's currently no native support for dot notation nesting in the groupBy query parameter

**Reply by martin-braun** [reply]
@rijkvanzanten I kinda worked around the problem, but my solution is much slower, but it works. About the backlog issue, please work on a solution so the community can still respond / react to rejected feature requests without causing clutter in the backlog. 

### It sends the wrong message

By simply locking the discussions you kinda give the message that new feature requests or solutions to the problems aren't welcome. On top of that, it feels like a clap in our face that we can't even communicate to each other on how to deal with exotic problems that the team cannot fix.

I now understand that you have different reasons to lock these discussions, because it makes your work easier, but then maybe GitHub Discussions aren't the right tool for the job.

### Please give as better ways to discuss

Please give us any platform or solution to communicate with each other in a forum style manner, and make it mandatory to use for feature requests, while it works with your backlogs and doesn't require you to lock discussions to de-clutter them. Or maybe tags could help to handle in hiding stale discussions, instead?

I tried to ask for help on your Discord, but it's clear that the helping community is overwhelmed by the frequency of requests over there. GitHub Discussions or any forum software allows a slower pace discussion and prevents good comments / solutions from being lost, as they can be found through Google.

### I like Directus, but it's not all blue sky

I like Directus, a lot! It's my absolute dream backend. When I need a new backend I instantly think about Directus and how I can make it work for my use-case. It's so frustrating that there are two channels of communication (GitHub Discussions and Discord) and you either cannot participate or your messages are getting lost. This is the biggest downside of Directus in my eyes. I cannot really connect in a meaningful way and help others who also struggle with similar exotic problems.

I mean ... look how I have to express my concerns to you in the first place. I need to abuse a discussion that is unrelated.

**I really hope you can do something about it.**

**Reply by rijkvanzanten** [reply]
Thanks for the insights 🙂 I agree; Discord has a lot of limitations and we've outgrown it's usefulness. Same with GitHub discussions unfortunately.. The current processes are shaped around the platforms we used, rather than finding the platforms to fit the ideal processes. That'll change soon though! I already had some ideas in flight to replace it with a Discourse instance, and your feedback was the final push I needed 🙂 We're just puzzling through the hosting, plugins, and configuration, but expect a new and improved community setup that hopefully take away these pains 🚀

**Comment by delivey** [comment]
Any updates?

**Comment by PhungVanHoa** [comment]
any news?

**Comment by EpariNigam** [comment]
Any update on this?

**Comment by noblica** [comment]
Regarding DISTINCT, you can kinda implement it using `aggregate` and `groupBy` (at least that's what I did). I'm using the Directus SDK, but you can replicate it through the API as well for sure

So you should be able to implement this:
```
SELECT DISTINCT country FROM customers
```

Like this this:
```
directusClient
    .request(
      aggregate("customers", {
        aggregate: {
          countDistinct: ["country"],
        },
        groupBy: ["country"],
      }),
    )
    .then((data) => data.map((item) => item.country));
```

**Comment by kilobyte2007** [comment]
@rijkvanzanten hi, was this feature planned?

**Reply by rijkvanzanten** [reply]
It's something we'd like to support in the future, but it's not currently in active development no
